stdlib/public/core/Stride.swift

```diff
@@ -237,7 +318,16 @@ extension StrideToIterator: IteratorProtocol {
     if _stride > 0 ? result >= _end : result <= _end {
       return nil
     }
-    _current = Element._step(after: _current, from: _start, by: _stride)
+    // The following check is needed because we might otherwise advance
+    // `_current` past the representable bounds of the `Strideable` type
+    // unnecessarily. Where a runtime error is not possible in such cases,
+    // `Strideable._step` returns a non-nil `index`.
+    if _current.index == nil
+        && result.distance(to: _end).magnitude < _stride.magnitude {
+      _current = (index: nil, value: _end)
```

Comment:
> I suspect this extra code might be responsible for the microbenchmark regression.

Note that the observed regression in the performance of `ArrayAppendSequence` shown in the first set of benchmarks pertained to a previous iteration of this PR, where we were checking `index == nil || index == 0`. It's not a great idea to do this for floating-point types, as I noted above.

As it turns out, checking `index == nil` doesn't cause a detectable benchmark regression except at `-Onone` (compare the second-last to last set of benchmarks). Although I haven't checked myself, I suspect that the compiler can see that `index` stays `nil` on each iteration and elides the whole thing out.

> A far worse problem is that distance(to:) may itself overflow, so this is just exchanging one set of trapping conditions with another set. 😭

You make an interesting point. This is one of those cases where it could be messier in _practice_ than in theory.

`Stride` represents the distance between two values of the `Strideable` type, and we do strongly imply (although for some reason do not explicitly state in any public-facing documentation) that `Stride` should be able to represent the distance between _any_ two values. That is, `distance(to:)` should _not_ overflow, and indeed it does not for the majority of types we vend. Where we botch this in the standard library for `Int` and some of the other integer types, it's noted in a "FIXME" that the choice of `Stride` type is incorrect.

If `Strideable` did not have such a requirement for `Stride` (with exceptions such as `Int` regarded as not-quite-right), then it would be kosher to have `struct Countdown: Strideable { typealias Stride = Int1 /* ... */ }` if we only ever wanted to stride backward by 1. Meanwhile, to align `Int1` with our other integer types, we'd probably write `struct Int1: Strideable { typealias Stride = Int /* ... */ }`. There does not appear to be _any_ useful algorithm generic over `Strideable` which would accommodate both of these types.

Put another way, `distance(to:)` may trap where a conforming type has chosen a suboptimal `Stride` type, and it reflects limitations in how a particular type has been made to conform to `Strideable`. By contrast, `advanced(by:)` may trap for _any_ type with a finite range of values (and, of course, it's perfectly legitimate to design types with a finite range of values) and such trapping is inherent to the design of this protocol requirement (i.e., `advanced(by:)` must return a value or else not return at all).

In either case, the trapping condition can be worked around for the purposes of strides by customizing `_step`. Overall, I'd argue that it's more elegant to require customizing `_step` for a conforming type that needs to work around a problem with its own conformance (i.e., where `distance(to:)` may trap because of the choice of `Stride` type) than to require such workarounds because of how the protocol itself has been designed through no fault of the concrete type that now has to do extra work.

_That said,_ you make an excellent point that this change in behavior could cause previously working code to trap; this may be _practically_ unacceptable if indeed third-party types implementing `Strideable` tend to do what you outline. I don't _think_ that happens in general, but I don't have concrete data to back that up.